crates/bevy_animation/src/lib.rs

```diff
@@ -397,196 +493,226 @@ struct AnimationTargetContext<'a> {
+                ref blend_weights,
+                ref mut schedule,
+                ..
+            } = *player;
 
-        // Advance transition animations.
-        player.transitions.retain_mut(|transition| {
-            // Decrease weight. Expire the transition if necessary.
-            transition.current_weight -= transition.weight_decline_per_sec * time.delta_seconds();
-            if transition.current_weight <= 0.0 {
-                return false;
-            }
+            // Tick animations, and schedule them.
 
-            if let Some(animation_clip) = animation_clips.get(&transition.animation.animation_clip)
-            {
-                transition
-                    .animation
-                    .update(time.delta_seconds(), animation_clip.duration);
-            };
+            schedule.clear();
 
-            true
+            let mut dfs = animation_graph.dfs();
+            let mut weights = HashMap::with_capacity(animation_graph.graph.node_count());
+
+            while let Some(node_index) = dfs.next(&animation_graph.graph) {
+                let node = &animation_graph[node_index];
+
+                let mut weight = node.weight;
+                for parent in animation_graph
+                    .graph
+                    .neighbors_directed(node_index, Direction::Incoming)
+                {
+                    if let Some(parent_weight) = weights.get(&parent) {
+                        weight *= parent_weight;
+                    }
+                }
+
+                if let Some(playing_animation) = playing_animations.get_mut(&node_index) {
+                    // Tick the animation if necessary.
+                    if !playing_animation.paused {
+                        if let Some(ref clip_handle) = node.clip {
+                            if let Some(clip) = animation_clips.get(clip_handle) {
+                                playing_animation.update(delta_seconds, clip.duration);
+                            }
+                        }
+                    }
+
+                    weight *= playing_animation.weight;
+                } else if let Some(&blend_weight) = blend_weights.get(&node_index) {
+                    weight *= blend_weight;
+                }
+
+                weights.insert(node_index, weight);
+
+                if let (Some(clip), Some(playing_animation)) =
+                    (&node.clip, playing_animations.get(&node_index))
+                {
+                    schedule.push(ScheduledAnimationClip {
+                        clip: clip.clone(),
+                        seek_time: playing_animation.seek_time,
+                        weight,
+                    });
+                }
+            }
+
+            // Sort animations by weight, so that the most heavily-weighted
+            // animations come first.
+            //
+            // FIXME: This seems dubious, but it matches Bevy 0.13 behavior. Figure
+            // out what we want to do here.
```

Comment:
I believe the RFC stated that the final weights in the GraphInfluences, or the ScheduledAnimationClips implemented here needed to have all of its weights normalized.

Reply:
I tried this, but it broke transitions. The problem is that, with the way transitions work in Bevy, when you transition to a new animation, the weight of the new animation immediately becomes 1.0, while the new animation's weight fades from 1.0 to 0.0. That means that, if you normalize, on the first frame the current animation's weight instantly drops from 1.0 to 0.5 normalize(1.0 + 1.0) == 0.5, which looks awful.

I think the normalization is wrong, but I'm not sure what the best thing to do here is.

Reply:
We may want to change this so that the new animation ramps up from 0.0 to 1.0 in tandem with the ramp down of the old animation. Unless there is some other factor, the weights should always add up to 1.0.

Reply:
I think it's still wrong if you did the normalization per bone too. You might have two animations, one targeting the Z rotation of a character's neck, and one targeting the X rotation. Normalizing them both to 0.5 would be wrong; you need them both to be 1.0.

I think normalization is just wrong no matter what.

Reply:
OK, I've implemented normalization per bone. This should work fine for now, and we can switch to Unity's semantics later (perhaps with an enhanced version of the `Animatable` trait).